src/data/sym/card.lean

```diff
@@ -5,41 +5,153 @@ Authors: Yaël Dillies, Bhavik Mehta
+    { unfold encode,
+      split_ifs,
+      { obtain ⟨a, _, ha⟩ := multiset.mem_map.mp h,
+        exact option.some_ne_none _ ha },
+      { refine map_injective (option.some_injective _) _ _,
+        convert eq.trans _ (decode α n (inl s)).attach_map_coe,
+        simp } },
+    { exact (dif_pos $ mem_cons_self _ _).trans (congr_arg _ $ erase_cons_head s _ _) }
+  end }
+
+end
+
+/-- `multichoose n k` is the number of multisets of cardinality `k` from a type of cardinality `n`.
+That is, it's the number of ways to select `k` items (up to permutation) from `n` items
+with replacement.
+
+This is defined as `nat.choose (n + k - 1) k`. It is related to the cardinality of `sym` in
+`sym.multichoose_eq`. -/
+def multichoose (n k : ℕ) := (n + k - 1).choose k
+
+lemma card_sym_rec (n : ℕ)
+  [fintype (sym (option α) n.succ)] [fintype (sym α n.succ)] [fintype (sym (option α) n)] :
+  card (sym (option α) n.succ) = card (sym α n.succ) + card (sym (option α) n) :=
+by { classical, simpa only [card_sum.symm] using fintype.card_congr (option_succ_equiv α n) }
+
+lemma multichoose_rec (n k : ℕ) :
+  multichoose n.succ k.succ = multichoose n k.succ + multichoose n.succ k :=
+by simp [multichoose, nat.choose_succ_succ, nat.add_comm, nat.add_succ]
+
+lemma multichoose_eq (α : Type*) [hα : fintype α] (k : ℕ) [fintype (sym α k)] :
+  multichoose (card α) k = card (sym α k) :=
+begin
+  classical,
+  generalize hn : card α + k = n,
+  unfreezingI { induction n with n ih generalizing α k },
+  { unfreezingI { obtain ⟨hn, rfl⟩ := add_eq_zero_iff.mp hn },
+    simp [multichoose, hn], },
+  { have : 0 < card α + k := by convert nat.succ_pos',
+    unfreezingI { cases k },
+    { haveI hne : nonempty α := card_pos_iff.mp this,
+      simp [multichoose], },
+    { obtain (hi|hi) := is_empty_or_nonempty α; haveI := hi,
+      { simp [multichoose, fintype.card_eq_zero], },
+      { haveI : inhabited α := classical.inhabited_of_nonempty (by assumption),
+        obtain ⟨β, βeqv⟩ := equiv.sigma_equiv_option_of_nonempty α,
+        haveI : fintype β := fintype_of_equiv_option βeqv.symm,
+        have βc : card β + k + 1 = n,
+        { simpa [fintype.card_congr βeqv,
+            nat.add_succ, nat.succ_add, add_zero, card_option] using hn },
+        have ih' := ih _ (k + 1) βc,
+        transitivity card (sym (option β) k.succ),
+        { rw [card_sym_rec, ← ih', fintype.card_congr βeqv, card_option, multichoose_rec],
+          have ih'' := ih α k,
+          rw [fintype.card_congr βeqv, card_option] at ih'',
+          specialize ih'' (by simpa [add_comm, add_assoc, add_left_comm] using βc),
+          rw [add_right_inj, ih''],
+          apply fintype.card_congr (sym.equiv_congr βeqv), },
+        { apply fintype.card_congr (sym.equiv_congr βeqv.symm), } } } },
+end
+
+/-- The *stars and bars* lemma: the cardinality of `sym α k` is equal to
+`nat.choose (card α + k - 1) k`. -/
+lemma stars_and_bars {α : Type*} [fintype α] (k : ℕ) [fintype (sym α k)] :
```

Comment:
We should mention how this fits into the [twelvefold way](https://en.wikipedia.org/wiki/Twelvefold_way). The type `sym α k` represents functions `fin k -> α` modulo permutation of the domain, so it's in the "n-multisubsets" box -- this is why it's "stars and bars."

We are sort of cheating here because `sym α k` is defined to be `{s : multiset α // s.card = k}`, but over in `data.sym.basic` we could eventually have an equivalence between `sym α k` and the correct type of functions mod permutation.